pyro/infer/mcmc/util.py

```diff
@@ -382,7 +384,8 @@ def initialize_model(model, model_args=(), model_kwargs={}, transforms=None, max
     # No-op if model does not have any discrete latents.
     model = poutine.enum(config_enumerate(model),
                          first_available_dim=-1 - max_plate_nesting)
-    model_trace = poutine.trace(model).get_trace(*model_args, **model_kwargs)
+    prototype_model = poutine.trace(InitMessenger(init_strategy)(model))
+    model_trace = prototype_model.get_trace(*model_args, **model_kwargs)
```

Comment:
@fehiepsi could you please review changes in this file? The changes are needed to
1. avoid calling `.sample()` in constructing the prototype `model_trace`, and
2. avoid duplicating expensive initialization work by reusing that `model_trace` as the first trace in `_find_valid_initial_params()`.

Reply:
IIUC, then with e.g. `init_to_uniform` strategy, we still [call sample](https://github.com/pyro-ppl/pyro/blob/dev/pyro/infer/autoguide/initialization.py#L111) but won't for some other strategies? If so, the change looks great to me.

Reply:
That's correct, `init_to_uniform` still calls sample. Thanks for reviewing!